### ISSUE TITLE
[BUG] pydeseq2 v0.4.7 throws IndexError due to anndata awkward variable length array not observed with 0.4.4

### ISSUE THREAD

**Issue by dguin** [owkin/PyDESeq2#251]
Anndata error while running pydeseq 0.4.7 but not with pydeseq 0.4.4 on the same dataset. Full error trace in error.txt and counts and metadata csv are included and code is given below
[error.txt](https://github.com/owkin/PyDESeq2/files/14594223/error.txt)
[example_counts.csv](https://github.com/owkin/PyDESeq2/files/14594224/example_counts.csv)
[example_metadata.csv](https://github.com/owkin/PyDESeq2/files/14594226/example_metadata.csv)
.

```
import pandas as pd
from pydeseq2.dds import DeseqDataSet
from pydeseq2.ds import DeseqStats

counts = pd.read_csv('example_counts.csv', index_col=0)
metadata = pd.read_csv('example_metadata.csv', index_col=0)

dds = DeseqDataSet(counts=counts,
                   metadata=metadata,
                   design_factors=list(metadata.columns),
                   refit_cooks=True,
                   ref_level=['condition', 'control'],
                   quiet=False)
dds.deseq2()
```

This dataset runs fine with pydeseq2 0.4.4 and haven't noticed this error at all with any of the older versions.

**Comment by BorisMuzellec** [comment]
Hi @dguin, thanks for reporting this bug. It will be fixed when #253 is merged.

Meanwhile you can work from the [corresponding branch](https://github.com/owkin/PyDESeq2/tree/fix_mean_disp_func).

**Comment by dguin** [comment]
Hi @BorisMuzellec thanks for the quick response and the quick turnaround. Out of curiosity I am wondering what part of the code changed that causes the break in 0.4.7 but not in 0.4.4. Is it simply a logistical change or did the algorithm change between these 2 versions?
